modules/standard/DistributedBag.chpl

```diff
@@ -0,0 +1,1245 @@
+  have a maximum of 125,000 elements stolen at once.
+*/
+config param WORK_STEALING_MAX_MEMORY_MB : real = 1.0;
+/*
+  Determines whether or not we check if we are 'frozen'. These checks cause a
+  significant decline in performance, of about 33%. If the utility is not needed,
+  as it is more useful when reads are performed more often than writes, it may
+  be disabled for a raw performance boost. If the user attempts to freeze the
+  bag with this set to true, it merely returns false; all other checks are elided.
+*/
+config param BAG_NO_FREEZE = false;
+/*
+  The minimum number of elements a horizontal segment must have to become eligible
+  to be stolen from. This may be useful if some segments produce less elements than
+  others and should not be stolen from.
+*/
+config param WORK_STEALING_MINIMUM = 1;
+
+/*
+  A parallel-safe distributed multiset implementation that scales in terms of
+  both nodes and processors per node. This data structure is unordered and employs
+  its own work-stealing algorithm, and provides a means to obtain a privatized instance of
+  the data structure for maximized performance. This is an ideal backbone for a
+  means to distribute work across nodes.
+*/
+class DistributedBag : Collection {
+  pragma "no doc"
+  var targetLocDom : domain(1);
+  /*
+    The locales to allocate bags for and load balance across.
+  */
+  var targetLocales : [targetLocDom] locale;
+  pragma "no doc"
+  var pid : int = -1;
+
+  // Node-local fields
+  pragma "no doc"
+  var bag : Bag(eltType);
+  pragma "no doc"
+  var concurrentTasks : atomic int;
+  pragma "no doc"
+  var frozenState : atomic int;
+
+  proc DistributedBag(type eltType, targetLocales : [?targetLocDom] locale = Locales) {
+    bag = new Bag(eltType, this);
+    this.targetLocDom = targetLocDom;
+    this.targetLocales = targetLocales;
+    pid = _newPrivatizedClass(this);
+  }
+
+  pragma "no doc"
+  proc DistributedBag(other, pid, type eltType = other.eltType) {
+    bag = new Bag(eltType, this);
+    this.targetLocDom = other.targetLocDom;
+    this.targetLocales = other.targetLocales;
+    this.pid = pid;
+  }
+
+  pragma "no doc"
+  proc dsiPrivatize(pid) {
+    return new DistributedBag(this, pid);
+  }
+
+  pragma "no doc"
+  proc dsiGetPrivatizeData() {
+    return pid;
+  }
+
+  pragma "no doc"
+  inline proc getPrivatizedThis {
+    if this.locale == here {
+      return this;
+    }
+
+    return chpl_getPrivatizedCopy(this.type, pid);
+  }
+
+  /*
+    Obtains a privatized version of this instance. The privatized version is a
+    cloned instance that is allocated on this node, which is useful for eliding
+    communications generated from accessing instance fields. Using another node's
+    instance will significantly penalize performance by bounding overall throughput
+    on the communication between the two nodes.
+  */
+  proc getPrivatizedInstance() {
+    return getPrivatizedThis;
+  }

```

Comment:
Right, that's just because it's a class not a record. The privatized instance ID is currently stored within the class itself - meaning that if you had an object of that type, you'd have to do a GET to retrieve the privatized ID in order to get your local copy, which sortof defeats the purpose of privatization. The pattern of using a record containing {pointer, private id} I described above is a reasonable way to solve this problem. I'm not so sure that it makes sense to expect users to manually call getPrivatizedInstance.